### DISCUSSION TITLE
How to handle error response in Next.js 13 Server Component?

### DISCUSSION THREAD

**Discussion by gidgudgod** [vercel/next.js#50771] [App Router]
So I'm trying to fetch data that needs token header for authorization in Server Component ``page.tsx``. The data itself comes from external backend/API and it can response with different status code. If ``Unauthorized`` it will response with status ``401``, if ``Not Found`` it will response with status ``404``. Here I already provided a handle for each different error status:

```typescript
const getItems = async (): Promise<IGetItemsResponse> => {
  const res = await fetch(`${process.env.API_URL}items`, {
    headers: {
      Authorization: `Bearer ${cookies().get('access_token')?.value}`,
    },
  });

  if (res.status === 404) {
    notFound()
  }
  if (res.status === 401) {
    const response = NextResponse.next();
    response.cookies.delete('access_token');
    response.cookies.delete('expiry_token');
    return response;
  }
  if (!res.ok) {
    throw new Error('Something went wrong!');
  }
  return res.json();
};

export default async function Page() {
  const items = await getItems()
  return <ItemList initialData={items}/>;
}

```

When it response with ``404`` Next.js succeed showing the nearest ``not-found.tsx``. However the problem here is the ``401`` error, I don't know why Next.js won't delete the cookie as I already handle the ``401`` error response there. Instead deleting the cookie, Next.js just run the ``throw new Error('Something went wrong!');`` right away which lead to showing the nearest ``error.tsx``. And then in ``error.tsx`` component, it shows the right error message in development build, but in production build it shows this different error message:
[error message in production build][1]

  [1]: https://i.stack.imgur.com/raHBg.png

Been searching and debugging for hours, yet still don't know why it won't handle the error.
**Please is there any way to properly handle this different error status in Next.js 13 Server Component?**
I need to at least delete the token cookies if it responses ``401``.

Thanks.

**Comment by egiie** [comment]
+1

**Comment by hardiksachan** [comment]
+1

Hey @gidgudgod did you figure out an approach to handle this?

**Comment by DannyPark1** [comment]
related to this https://github.com/vercel/next.js/issues/48681#issuecomment-1633579198
we need this so bad

**Reply by kelvinbush** [reply]
+1

**Comment by lesimoes** [comment]
According to the docs when triggers a error on server component Next will look to nearest error file to render.

> If an error is thrown inside a Server Component, Next.js will forward an Error object (stripped of sensitive error information in production) to the nearest error.js file as the error prop.

https://nextjs.org/docs/app/building-your-application/routing/error-handling

You can use global-error.jsx (.tsx) to catch all errors

``` javascript
'use client';

export default function GlobalError({
  error,
  reset,
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  return (
    <html>
      <body>
        <h2>Something went wrong!</h2>
        <button onClick={() => reset()}>Try again</button>
      </body>
    </html>
  );
}

```

**Note**: This not works in dev mode, remember to run your application in prod mode to verify this functionality.

**Reply by saver711** [reply]
What if i need to show inline error
not a full error page
